feat(astro): add Built-in SVG component support

stramel opened the pull request:
## Changes

This PR changes how Astro handles SVG files by default following up on the discussions on https://github.com/withastro/roadmap/issues/699

By default, SVG files will be components and can be imported by referencing their file path and then treating them like a component.  Take this for example:

**previously:**
```astro
---
import logo from './path/to/svg/file.svg'
---

<div set:html={logo}  style={{height: '24px', width: '24px' }} />
```

**now:**

```astro
---
import Logo from './path/to/svg/file.svg';
---

<Logo size={24} />
```

This change currently only offers a foundation to build upon. 
- Automatic SVG Sprite with the option to directly embed/inline (`inline` prop currently) the SVG instead
- Provides the ability to be extended by third-parties
	- For instance using `astro-icon`, which has been the critical plugin when it comes to icons, as a playground, I was able to extend the foundation built here to add dynamic icon import for Iconify.

As far as SVG optimizations go, it only does some basic stripping of a few attributes. A future enhancement would be integrating SVGO or similar SVG optimizer.

/cc @natemoo-re 

## Testing

I have updated the blog example to include using the SVG components. I have also tested locally against `astro-icon` and its demo application.
<!-- How was this change tested? -->
<!-- DON'T DELETE THIS SECTION! If no tests added, explain why. -->

## Docs

This will definitely need documentation added to it. Will need some assistance in writing it.
<!-- Could this affect a user’s behavior? We probably need to update docs! -->
<!-- If docs will be needed or you’re not sure, uncomment the next line: -->
/cc @withastro/maintainers-docs for feedback!

<!-- DON'T DELETE THIS SECTION! If no docs added, explain why.-->
<!-- https://github.com/withastro/docs -->

natemoo-re commented on packages/astro/client.d.ts:
```diff
@@ -106,14 +107,31 @@ declare module '*.webp' {
+		 * Shorthand for setting the `height` and `width` properties
+		 */
+		size?: number | string;
+		/**
+		 * Bypasses automatic sprite optimization by directly inlinining the SVG
+		 */
+		inline?: boolean
```
After sitting with this for a while, inline behavior should be the default. There are a few edge-cases that have come up with dynamic client-side behavior in `astro-icon`, so while I still think the optimization is very useful, we should prioritize flexibility by keeping the sprite behavior opt-in.

stramel replied:
You're right. I think especially with the SVG imports that developers wouldn't necessarily assume the optimization by default. 

My first thought is to offer a "mode" that allows for "sprite" or "inline" ("inline" being the default) that is configurable in the Astro Config. I assume we'll later have some other properties under and "svg" scope such as an "image service" (like svgo).

I have also seen an interesting use-case (https://github.com/natemoo-re/astro-icon/pull/238) that we might be able to support later for exporting symbols instead. This would be fairly trivial with a "mode" and might enable some better DX around framework usages.

I would like to discuss and perhaps iterate a bit on the way we allow developers to opt-in/out of inline/sprite usages. Will bring this up in the Stage 3 RFC.

natemoo-re commented on packages/astro/src/assets/runtime.ts:
```diff
@@ -0,0 +1,53 @@
+	makeNonEnumerable(Component);
+
+	if (import.meta.env.DEV) {
+		Object.defineProperty(Component, Symbol.for('nodejs.util.inspect.custom'), {
+			value: (_: any, opts: any, inspect: any) => inspect(meta, opts),
+		});
```
Not sure how necessary this is, but maintaining the current `console.log` behavior seems like a good idea.

A comment here explaining why we're doing this would be helpful 👍

natemoo-re commented on packages/astro/src/assets/utils/node/emitAsset.ts:
```diff
@@ -42,6 +43,11 @@ export async function emitESMImage(
 		value: id,
 	});
 
+	// Attach file data for SVGs
+	if (fileMetadata.format === 'svg') {
+		emittedImage.contents = fileData;
+	}
```
Would love @Princesseuh's take on this. Any expected footguns with including the file content for SVGs in the image metadata?

Princesseuh replied:
If you have a lot of SVGs, this can be problematic for SSR since it'll get in your bundle, that's my only concerns

natemoo-re replied:
Right. Unfortunately that will be unavoidable for this feature.

I think the DX improvement is worth the tradeoff, though!

stramel replied:
This is definitely outside my realm of expertise. I'll defer to whatever the consensus is here :)

natemoo-re commented on packages/astro/src/assets/utils/svg.ts:
```diff
@@ -0,0 +1,54 @@
+}
+
+export function normalizeProps(attributes: SvgAttributes, { size, ...props }: SvgAttributes) {
+	if (size) {
+		props.height = size;
+		props.width = size;
+	}
```
Looks like `size` will override `width` and `height` attributes if present. I think `size` should only apply when they are `undefined`.

stramel replied:
Updated! Will add a test case for this as well!

natemoo-re commented on packages/astro/test/core-image-svg.test.js:
```diff
@@ -0,0 +1,264 @@
+					assert.equal($svg.attr('width'), '48');
+					assert.equal(!!$svg.attr('size'), false);
+				});
+				it('has height and width overridden - size set', () => {
+					let $svg = $('#override-attrs svg');
+					assert.equal($svg.length, 1);
+					assert.equal($svg.attr('height'), '16');
+					assert.equal($svg.attr('width'), '16');
+					assert.equal(!!$svg.attr('size'), false);
+				});
```
This test will need updating if the `size` behavior switches

natemoo-re commented on packages/astro/test/core-image-svg.test.js:
```diff
@@ -0,0 +1,264 @@
+					assert.equal(!!$svg.attr('version'), false);
+				});
+			});
+			describe('additional props', () => {
```
Would be nice to assert that the default `role="img"` is overwritable. Now that I think of it, we should also double check that `role="img"` is the best default since I know `aria-hidden` and `role="presentation"` are also used sometimes.

stramel replied:
Will check for some supporting material on this :) Added a test in the mean time to assert it's overridable

stramel replied:
Everything I'm finding still says `role="img"` for image SVGs paired with a `title` element and/or `aria-label`/`aria-labelledby`. 

For removing it from the accessibility tree, it is still recommended to use `aria-hidden=true`.

natemoo-re commented on packages/astro/test/fixtures/core-image-svg/src/assets/astro.svg:
😩 this is the old logo

stramel replied:
Oh no! Sorry about that! Just grabbed it from the logos svg library. Will find the updated and update here!

natemoo-re commented on packages/astro/src/assets/utils/svg.ts:
```diff
@@ -0,0 +1,54 @@
+import { parse, renderSync } from 'ultrahtml';
```
Just a callout that I maintain `ultrahtml`—this could easily be swapped for another XML parser if preferred, but keep in mind that it will be bundled into the server runtime so we should try to keep a light footprint.

natemoo-re reviewed (commented):
Left a few suggestions, but this looks really good!

stramel commented on packages/astro/client.d.ts:
```diff
@@ -1,6 +1,7 @@
 /// <reference types="vite/types/import-meta.d.ts" />
 /// <reference path="./types/content.d.ts" />
 /// <reference path="./types/actions.d.ts" />
+/// <reference path="./jsx-runtime.d.ts" />
```
It might just be my development machine. I was having some issues with the devcontainers on windows. 

It wasn't showing the `astroHTML` as a valid option until I added that. Doesn't seem to be an issue anymore. I'll remove now, thank you!

stramel commented on examples/blog/src/components/Footer.astro:
```diff
@@ -7,35 +11,15 @@ const today = new Date();
-					d="M11.19 12.195c2.016-.24 3.77-1.475 3.99-2.603.348-1.778.32-4.339.32-4.339 0-3.47-2.286-4.488-2.286-4.488C12.062.238 10.083.017 8.027 0h-.05C5.92.017 3.942.238 2.79.765c0 0-2.285 1.017-2.285 4.488l-.002.662c-.004.64-.007 1.35.011 2.091.083 3.394.626 6.74 3.78 7.57 1.454.383 2.703.463 3.709.408 1.823-.1 2.847-.647 2.847-.647l-.06-1.317s-1.303.41-2.767.36c-1.45-.05-2.98-.156-3.215-1.928a3.61 ...
-				></path></svg
-			>
+			<MastodonLogo size={32} aria-hidden="true" />
```
Sounds good I'll drop that commit :)

Princesseuh commented on packages/astro/src/assets/utils/svg.ts:
```diff
@@ -0,0 +1,54 @@
+type SvgAttributes = Record<string, any>;
+
+/**
+ * Some attributes required for `image/svg+xml` are irrelevant when inlined ina `text/html` document. We can save a few bytes by dropping them.
```
```suggestion
 * Some attributes required for `image/svg+xml` are irrelevant when inlined in a `text/html` document. We can save a few bytes by dropping them.
```

stramel replied:
Nice catch! Thank you!

Princesseuh commented on packages/astro/src/assets/vite-plugin-assets.ts:
```diff
@@ -229,6 +230,11 @@ export default function assets({
 
+					if (/\.svg$/.test(id)) {
+						const { contents, ...metadata } = imageMetadata;
+						return makeSvgComponent(metadata, contents!);
```
A comment explaining why `contents` is always defined would be great, you can also cast `imageMetadata` to avoid the `!` (still with a comment explaining why it gets casted)

stramel commented:
I have added it behind an experiment flag called `svg`. With this change, I swapped the default implementation to be a more straightforward approach of inlining the SVG. In order to control this a bit better, I have a `mode` setting in the `svg` config to apply a default, which is defaulted to `inline`. I also changed the `inline` prop to a `mode` prop that will allow configuring between `inline` and `sprite` on specific SVGs.

ematipico commented:
Thank you, @stramel, for kicking off the feature. The core discussed the feature a little bit, and we felt to land this feature for Astro v5.

Now that we are in a beta phase, the v4 branch will accept only small features; plus, SVG built-in support is quite big, so we want to have enough time to craft a good blog post. What do you think?

You'd have to point your changes to `next` instead of `main`. It's possible that the code changed in the meantime, so let us know if you need any help.

stramel commented:
Sounds great, I'll update this PR! Thanks for taking the time to discuss it.

stramel commented:
@ematipico @matthewp Sorry for the delay! I got the PR switched over to target the `next` branch now.

stramel commented:
Going to double-check everything tomorrow! Hopefully we're good to merge after that

stramel commented:
Should be good to go!

sarah11918 commented:
Just pointing out I haven't seen any docs for this yet, so please let me know if those are coming or if help is needed!

stramel commented:
> Just pointing out I haven't seen any docs for this yet, so please let me know if those are coming or if help is needed!

I'm not familiar with the docs process. I would be happy to write some if you could point me to a previous PR or something as a guide. I also would welcome some help as writing isn't my forte 🙂

ematipico commented:
@stramel 

Since we are about to send this PR to `next`, you'll have to send the docs PR to their respective branch, which is [`5.0.0-beta`](https://github.com/withastro/docs/tree/5.0.0-beta).

As for where, I would start with the Images guide: https://docs.astro.build/en/guides/images/ 

https://github.com/withastro/docs/blob/5.0.0-beta/src/content/docs/en/guides/images.mdx

As for **where** in the page, I don't exactly, but I would start with a new paragraph (maybe an h2 heading) and send the PR, the Sarah and team will take it from there, and will help you to craft the content.

bluwy commented on packages/astro/src/assets/runtime.ts:
```diff
@@ -0,0 +1,54 @@
+	unescapeHTML,
+} from '../runtime/server/index.js';
+import type { ImageMetadata } from './types.js';
+import { makeNonEnumerable, normalizeProps } from './utils/svg.js';
```
Can we inline both of these function into this file? It looks to only be used here. That way we don't import `svg.js` which may bundle `ultrahtml` into the SSR runtime, or in the future may accidentally bundle other things in that file.

bluwy commented on packages/astro/src/assets/runtime.ts:
```diff
@@ -0,0 +1,54 @@
+		return render`<svg${spreadAttributes({viewBox, ...normalizedProps})}>${title}${unescapeHTML(children)}</svg>`
+	});
+
+	makeNonEnumerable(Component);
```
Is there a reason to make non-enumerable here?

bluwy replied:
Looks like it's to prevent revealing that this is a component. I think I'm fine with it, though I wonder if we could make this dev only

bluwy commented on packages/astro/src/assets/runtime.ts:
```diff
@@ -0,0 +1,54 @@
+	children: string;
+}
+
+/**
+ * Make sure these IDs are kept on the module-level so they're incremented on a per-page basis
+ */
+let ids = 0;
```
IIUC, if this is on the module-level (and also globally as a singleton), wouldn't this value not be incremented on a per-page basis? If per page, I think it should be bound to the `result` from `createComponent` as a single `result` is created when rendering per page. So maybe we can use a WeakMap here to track the ids per `result`, which we can also use it to track the first render instead of using `WeakSet<Response>`?

stramel replied:
I'm not quite following how to use the `WeakMap` here.

bluwy replied:
The suggestion here is to use `const ids = new WeakMap<SSRResult, number>` instead to track the `id` per `result` (from `createComponent`)

stramel replied:
I'm having trouble getting this to work. I'll give it a shot again tomorrow. Maybe I'm just tired 😅

ematipico replied:
I'll help you out :)

ematipico replied:
Here [`6ab0f1a` (#12067)](https://github.com/withastro/astro/pull/12067/commits/6ab0f1acc2b0792845ff9e75f0bca022f1593a84)

bluwy replied:
We don't need the `counter` variable there. What I mean is that the WeakMap values are the counter itself that starts from 0 on first assigned.

```js
if (!ids.has(result)) ids.set(result, 0)
const id = ids.get(result)
ids.set(result, id + 1)
```

stramel replied:
I'll give it a shot but I was having issues previously with this.

bluwy commented on packages/astro/src/assets/vite-plugin-assets.ts:
```diff
@@ -213,6 +214,12 @@ export default function assets({ settings }: { settings: AstroSettings }): vite.
 						});
 					}
 
+					if (settings.config.experimental.svg && /\.svg$/.test(id)) {
+						const { contents, ...metadata } = imageMetadata;
+						// We know that the contents are present, as we only emit this property for SVG files
+						return makeSvgComponent(metadata, contents!, { mode: settings.config.experimental.svg.mode });
+					}
```
Is the intent of the if condition that `settings.config.experimental.svg` must be set, e.g. `{}`, to opt in to this feature? Or is the feature will always be enabled?

stramel replied:
yeah, this wasn't quite sure how to handle this. Wanted it to be easy to opt into but wanted options as well. so i wanted a `boolean | { mode: 'inline' | 'sprite' }` but wasn't sure how to handle that.

bluwy replied:
That types looks good to me. You can declare it like this: https://github.com/stramel/astro/blob/fcc6eca76e427e50f39263ebdd579bdecb1f573c/packages/astro/src/core/config/schema.ts#L238-L246

And handle that the types could be a boolean or object. You can normalized it to an object with `.transform()` in the schema if you'd like.

natemoo-re commented on .changeset/blue-socks-doubt.md:
````diff
@@ -0,0 +1,15 @@
+'astro': minor
+---
+
+Add built-in SVG components
+
+SVGs can now be utilized as components for easier inlining, eg:
+
+```astro
+---
+import Logo from './path/to/svg/file.svg';
+---
+
+<Logo size={24} />
+```
````
Maybe a bit more detail?

```suggestion
Adds experimental support for built-in SVG components.

After enabling the `experimental.svg` flag, `.svg` files can be imported and used as components. They will be inlined into the HTML output.
```

natemoo-re reviewed (commented):
Looks really good!

ematipico commented:
@stramel While I was working on the PR, I noticed that you added a new configuration in Astro, which is fine, however the RFC doesn't mention anything about it. We should update the RFC and explain the options. 

Also, here's some feedback regarding experimental options. There aren't written rules, but in the past I was in your same situation and I want for this route:

```ts
export default defineConfig({
	svg: {
		mode: "inline"
	},
	experimental: {
		svg: true
	}
})
```

We still need to have the experimental flag, which stays a boolean every time. Then, you create a new section in the Astro configuration. You're still free to change it as you see fit, because it's part of the experimental feature, however I found that the maintenance of these two separate is easier for us maintainers.

sarah11918 commented on .changeset/blue-socks-doubt.md:
````diff
@@ -0,0 +1,17 @@
+
+Adds experimental support for built-in SVG components.
+
+After enabling the `experimental.svg` flag, `.svg` files can be imported and used as components. They will be inlined into the HTML output.
+
+```astro
+---
+import Logo from './path/to/svg/file.svg';
+---
+
+<Logo size={24} />
+```
+
+To learn more, check out [the documentation](https://docs.astro.build/reference/configuration-reference/#experimentalsvg).
````
``````suggestion

This feature allows you to import SVG files directly into your Astro project as components. By default, Astro will inline the SVG content into your HTML output.

To enable this feature, set `experimental.svg` to `true` in your Astro config:

```js
{
  experimental: {
    svg: true,
  },
}
```

To use this feature, import an SVG file in your Astro project, passing any common SVG attributes to the imported component. Astro also provides a `size` attribute to set equal `height` and `width` properties:

```astro
---
import Logo from './path/to/svg/file.svg';
---

<Logo size={24} />
```

For a complete overview, and to give feedback on this experimental API, see the [Feature RFC](https://github.com/withastro/roadmap/pull/1035).
`````` 

For a new feature, we do like to hype it up a bit! The changeset is a great place to show both the config flag needed and usage. I think something like this works!

sarah11918 commented on packages/astro/src/types/public/config.ts:
````diff
@@ -1699,6 +1700,66 @@ export interface ViteUserConfig extends OriginalViteUserConfig {
+		 * }
+		 * ```
+		 * 
+		 * To use this feature, import an SVG file in your Astro project:
````
```suggestion
		 * To use this feature, import an SVG file in your Astro project, passing any common SVG attributes to the imported component.
		 * Astro also provides a `size` attribute to set equal `height` and `width` properties:
```
Matching the changeset, and providing a little more usage guidance that people should be able to run with quickly.

sarah11918 commented on packages/astro/src/types/public/config.ts:
```diff
@@ -1699,6 +1700,66 @@ export interface ViteUserConfig extends OriginalViteUserConfig {
+			 * @type {string}
+			 * @default 'inline'
+			 * 
+			 * This mode will determine how Astro handles the imported SVG file by default. Astro will inline the SVG content into your HTML output if not specified.
```
```suggestion
			 * The default technique for handling imported SVG files. Astro will inline the SVG content into your HTML output if not specified.
```
In my searching for content about this topic, I saw the word "technique" used a lot when comparing inline vs SVG sprites, so if you think that's a common phrase, let's go with something like that? It's nice to use a vocab familiar to the topic!

Also, we would normally show setting the config here, though it looks like it's undecided whether this will be `experimental.svg.mode` or `svg.mode`. 

I guess the maintainers prefer you make a config option outside of the experimental flag (though docs always finds it easier when the config is with the flag, because we don't have to show config in two different places, and figure out how to document a "normal" property that only exists if you've enabled a flag.... Fight, fight, fight!  :joy: )

stramel replied:
I think I prefer it in the `experimental` section so there is less confusion around having to also turn on the `experimental.svg` flag if someone just sees the `svg` section. 

I had made some changes to better handle the `experimental.svg` as a `boolean` and normalized it into the object. I think this will move well once stable too.

But I'm happy to do whatever the consensus is :)

sarah11918 commented on packages/astro/src/types/public/config.ts:
```diff
@@ -1699,6 +1700,66 @@ export interface ViteUserConfig extends OriginalViteUserConfig {
+			 * - `inline`: Astro will inline the SVG content into your HTML output.
+			 * - `sprite`: Astro will generate a sprite sheet with all imported SVG files.
+			 * 
+			 * When using 'sprite' mode, it is important to note that the symbol definition will be included with the first instance of the SVG that is rendered. This means that if the first instance is hidden or removed from the DOM, all other references will break.
```
```suggestion
			 * When using 'sprite' mode, it is important to note that the symbol definition will be included with the first instance of the SVG that is rendered. This means that if the first instance is hidden or removed from the DOM, all other references will break.
```
Question: is this unique to Astro's implementation, or would this be something common to using a sprite sheet in general? I get that it's a caveat, but if it's not an *Astro-specific* caveat then it's a little bit more for the reader to have to process before they even see how to use this feature.

I might suggest we either remove this paragraph entirely (we do say full details in the RFC!), or place it after the code example if this is a super key thing and people will have a lot of trouble using the feature if they don't know this immediately up front.

stramel replied:
This is unique to the implementation that we chose to go with.  Though it would probably apply to the spritesheet in general but is significantly less common if you just have a single spritesheet. 

This is something I was debating about adding as an enhancement later or opt-in.  Would need some help from the core team on it though. 

I'll leave it up to you if we keep this caveat or remove it.

sarah11918 replied:
OK, for this stage of the game, I say we omit the caveat in these docs here. We just get people up and running quickly and happily, and they can check the RFC when and if they run into trouble!

sarah11918 commented on packages/astro/src/types/public/config.ts:
```diff
@@ -1699,6 +1700,66 @@ export interface ViteUserConfig extends OriginalViteUserConfig {
+			 * import Logo from './path/to/svg/file.svg';
+			 * ---
+			 * 
+			 * <Logo size={24} mode="inline" />
```
```suggestion
			 * <Logo size={24} mode="sprite" />
```
Since `inline` is the default, it's maybe more common that someone would be setting `sprite` to override a default `inline` configuration. So, we'd usually show setting something not the default. (Granted, they could have set default to `sprite` and need to overwrite that! But this feels like the more common thing to show.)

sarah11918 reviewed (commented):
Thank you so much for this amazing feature @stramel ! I think people are going to really enjoy this!  :raised_hands: 

I've done a quick pass on the documentation (changeset and flag docs) and you'll find some suggestions below for your consideration!

I know Nate has a docs PR going, and it seems like the `mode` property is newer than that, so we'll just make sure that we've got everything we need in the docs PR, too, before docs totally signs off on this!  :rocket:

sarah11918 commented on packages/astro/src/types/public/config.ts:
```diff
@@ -1699,6 +1700,67 @@ export interface ViteUserConfig extends OriginalViteUserConfig {
+			 * 
+			 * - `inline`: Astro will inline the SVG content into your HTML output.
+			 * - `sprite`: Astro will generate a sprite sheet with all imported SVG files.
+			 * 
+			 * When using 'sprite' mode, it is important to note that the symbol definition will be included with the first instance of the SVG that is rendered. This means that if the first instance is hidden or removed from the DOM, all other references will break.
```
```suggestion
```
Removing the caveat here as discussed!

sarah11918 reviewed (approved):
Approving for docs!

(Also noting, as I always do, that I don't see any new errors added with this feature. If that changes before the feature is released, then please loop me back in!  :raised_hands: )

stramel commented:
Thank you all for the help to get this in! I greatly appreciate all the reviews and feedback!

stramel commented:
Seems there is some kind of issue with the refactor of `ids` to use `WeakMap`. I'm just getting all the same icon from the integration with `astro-icon` through the `makeSvgComponent`. I'll open a new PR for it

stramel commented:
Initial fix: https://github.com/withastro/astro/pull/12511

Going to do some testing and ensure I add a test to fail on previous case. Tested with `astro-icon`
